library/alloc/src/string.rs

```diff
@@ -117,27 +117,84 @@ use crate::vec::Vec;
 ///
 /// # UTF-8
 ///
-/// `String`s are always valid UTF-8. This has a few implications, the first of
-/// which is that if you need a non-UTF-8 string, consider [`OsString`]. It is
-/// similar, but without the UTF-8 constraint. The second implication is that
-/// you cannot index into a `String`:
+/// `String`s are always valid UTF-8. If you need a non-UTF-8 string, consider
+/// [`OsString`]. It is similar, but without the UTF-8 constraint. Because UTF-8
+/// is a variable width encoding, `String`s can be smaller than an array of
+/// the same `chars`:
+///
+/// ```
+/// use std::mem;
+///
+/// // `s` is ASCII which represents each `char` as one byte
+/// let s = "hello";
+/// assert_eq!(s.len(), 5);
+///
+/// // A `char` array with the same contents would be longer because
+/// // every `char` is four bytes
+/// let s = ['h', 'e', 'l', 'l', 'o'];
+/// let size: usize = s.into_iter().map(|c| mem::size_of_val(&c)).sum();
+/// assert_eq!(size, 20);
+///
+/// // However, for non-ASCII strings, the difference will be smaller
+/// // and sometimes they are the same
+/// let s = "💖💖💖💖💖";
+/// assert_eq!(s.len(), 20);
+///
+/// let s = ['💖', '💖', '💖', '💖', '💖'];
+/// let size: usize = s.into_iter().map(|c| mem::size_of_val(&c)).sum();
```

Comment:
The size_of_val is a little weird here, but I guess OK -- technically it can just be size_of::<char>(). That might be a little more obvious, not sure; it feels like the key point is that `Vec<char>` uses a constant 4 bytes per codepoint (unicode scalar value) whereas UTF-8 uses between 1-4 bytes; that means that it's never worse than `Vec<u32>` but sometimes (quite rarely!) equivalent.

Reply:
Yeah I figured `size_of_val` would more semantically convey "the size of the contents" where `size_of::<char>()` ignored the `c` so it didn't seem as obvious to me what I was measuring. But your key point is spot on - if you store each `char` you will probably (but won't always) get worse compression than if you store them in a `String`.